server/keyshare/tasks/tasks.go

```diff
@@ -0,0 +1,157 @@
+	if err != nil {
+		t.conf.Logger.WithField("error", err).Error("Could not remove email login tokens that have expired")
+		return
+	}
+	_, err = t.db.Exec("DELETE FROM irma.email_verification_tokens WHERE expiry < $1", time.Now().Unix())
+	if err != nil {
+		t.conf.Logger.WithField("error", err).Error("Could not remove email verification tokens that have expired")
+	}
+}
+
+// Cleanup accounts disabled long enough ago.
+func (t *taskHandler) cleanupAccounts() {
+	_, err := t.db.Exec("DELETE FROM irma.users WHERE delete_on < $1 AND (coredata IS NULL OR last_seen < delete_on - $2)",
+		time.Now().Unix(),
+		t.conf.DeleteDelay*24*60*60)
+	if err != nil {
+		t.conf.Logger.WithField("error", err).Error("Could not remove accounts scheduled for deletion")
+	}
+}
+
+func (t *taskHandler) sendExpiryEmails(id int64, username, lang string) error {
+	// Fetch user's email addresses
+	err := t.db.QueryIterate("SELECT email FROM irma.emails WHERE user_id = $1",
+		func(emailRes *sql.Rows) error {
+			var email string
+			err := emailRes.Scan(&email)
+			if err != nil {
+				return err
+			}
+
+			// And send
+			err = t.conf.SendEmail(
+				t.conf.deleteExpiredAccountTemplate,
+				t.conf.DeleteExpiredAccountSubjects,
+				map[string]string{"Username": username, "Email": email},
+				[]string{email},
+				lang,
+			)
+			return err
+		},
+		id,
+	)
+	if err != nil {
+		t.conf.Logger.WithField("error", err).Error("Could not retrieve user's email addresses")
+		return err
+	}
+	return nil
+}
+
+// Mark old unused accounts for deletion, and inform their owners.
+func (t *taskHandler) expireAccounts() {
+	// Disable this task when email server is not given
+	if t.conf.EmailServer == "" {
+		t.conf.Logger.Warning("Expiring accounts is disabled, as no email server is configured")
+		return
+	}
+
+	// Iterate over users we havent seen in ExpiryDelay days, and which have a registered email.
+	// We ignore (and thus keep alive) accounts without email addresses, as we can't inform their owners.
+	// (Note that for such accounts we store no email addresses, i.e. no personal data whatsoever.)
+	err := t.db.QueryIterate(`
+		SELECT id, username, language
+		FROM irma.users
+		WHERE last_seen < $1 AND (
+			SELECT count(*)
+			FROM irma.emails
+			WHERE irma.users.id = irma.emails.user_id
+		) > 0
+		LIMIT 10`,
+		func(res *sql.Rows) error {
+			var id int64
+			var username string
+			var lang string
+			err := res.Scan(&id, &username, &lang)
+			if err != nil {
+				return err
+			}
+
+			// Send emails
+			err = t.sendExpiryEmails(id, username, lang)
+			if err != nil {
+				return err // already logged, just abort
+			}
+
+			// Finally, do marking for deletion
+			err = t.db.ExecUser("UPDATE irma.users SET delete_on = $2 WHERE id = $1", id,
+				time.Now().Add(time.Duration(24*t.conf.DeleteDelay)*time.Hour).Unix())
+			if err != nil {
+				return err
+			}
+			return nil
+		},
+		time.Now().Add(time.Duration(-24*t.conf.ExpiryDelay)*time.Hour).Unix(),
```

Comment:
Why not use `time.Now().Sub(...)` instead of using a negative duration. It puzzled me a bit.

Reply:
`time.Now().Sub(x)` is not the same as `time.Now().Add(-x)`: `Sub(x)` takes a `time.Time` and returns the difference between the two; and `Add` takes instead a `time.Duration` and adds that. So the one returns a `time.Duration` and the other a `time.Time`. As far as I know, the only way to subtract a negative time duration from a time is how it is done here.

Reply:
Ah funny difference, I wouldn't have concluded that when looking to the names. Then it is fine of course.